node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -369,51 +269,75 @@ async fn distribute_collation<Context>(
 	state: &mut State,
 	id: ParaId,
 	receipt: CandidateReceipt,
+	parent_head_data_hash: Hash,
 	pov: PoV,
 	result_sender: Option<oneshot::Sender<CollationSecondedSignal>>,
 ) -> Result<()> {
-	let relay_parent = receipt.descriptor.relay_parent;
+	let candidate_relay_parent = receipt.descriptor.relay_parent;
+	let candidate_hash = receipt.hash();
 
-	// This collation is not in the active-leaves set.
-	if !state.view.contains(&relay_parent) {
-		gum::warn!(
+	let per_relay_parent = match state.per_relay_parent.get_mut(&candidate_relay_parent) {
+		Some(per_relay_parent) => per_relay_parent,
+		None => {
+			gum::debug!(
+				target: LOG_TARGET,
+				para_id = %id,
+				candidate_relay_parent = %candidate_relay_parent,
+				candidate_hash = ?candidate_hash,
+				"Candidate relay parent is out of our view",
+			);
+			return Ok(())
+		},
+	};
+	let relay_parent_mode = per_relay_parent.prospective_parachains_mode;
+
+	let collations_limit = match relay_parent_mode {
+		ProspectiveParachainsMode::Disabled => 1,
+		ProspectiveParachainsMode::Enabled => MAX_CANDIDATE_DEPTH + 1,
```

Comment:
When async backing is disabled, the only possible depth is 0 (i.e. limit is 1)
When enabled it's 0 + 1..=MAX_CANDIDATE_DEPTH

Reply:
> Also a limit of 4 for candidates for a single relay parent seems excessive: Parachains progressing 4 times as fast as the relay chain - assuming we can back multiple candidates per block? For providing good asynchronicity 2 should already provide a good value.
> 
> As in collator is already done with providing one collation, before the next relay parent appeared - can already start building the next, which might be heavier - instead of wasting time waiting for the next relay parent for starting.

If a collator builds 4 (in fact, 5) consecutive parablocks and advertises it for one relay parent, it will not be able to advertise anything else on top of this chain because hypothetical depth request will tell us to reject it.

The way it works is: prospective parachains keeps a fragment tree per active leaf per parachain.
When a leaf is replaced with descendant, the candidate storage is transferred to a new leaf, if some
candidates leave the allowed ancestry threshold, they're pruned.

The only harm it can do is that collator spam validators with 5 different candidates built on top of e.g. the most recent para head. We could either restrict them from advertising the same para head multiple times, or track occupied depths
the same way it's done in backing.

cc @rphmeier